##### This Extracts the interaction effects genomic risk loci generated by FUMA
###   Creates a CSV with rsID, uniqueID, beta, pval, nearest gene, and marginal/ joint pvals 

# remove everything in environment
rm(list=ls())

print(getwd())

# This will create a csv file comparing Beta and P values between marginal and interaction results to joint results 
# Setting Paths

r.libs = "/projects/sunlab/R.lib"

# Load libraries

suppressPackageStartupMessages(library(data.table,lib.loc= r.libs))
suppressPackageStartupMessages(library(crayon,lib.loc= r.libs))
suppressPackageStartupMessages(library(dplyr,lib.loc= r.libs))
suppressPackageStartupMessages(library(cli,lib.loc= r.libs))

# Loading data

inter_loci = fread("FUMA_results/FUMA_inter/GenomicRiskLoci.txt")  
snp_info = fread("FUMA_results/FUMA_inter/snps.txt")
marg = fread("marginal_FUMA.txt")
joint = fread("joint_FUMA.txt")


# Subsetting DFs
inter_loci = inter_loci[,c("uniqID", "rsID")]  # isolating uniqueID and rsID
snp_info = snp_info[,c("rsID", "nearestGene", "beta", "gwasP")]
names(snp_info)[4] = "inter_Pval"

# merging DFs
inter_info = merge(inter_loci, snp_info, by = "rsID")

# adding in joint and marg p values
inter_info$joint_Pval = joint[joint$SNPID %in% inter_info$rsID,"P"]

inter_info$marg_Pval = marg[marg$SNPID %in% inter_info$rsID,"P"]

# Save results
write.csv(inter_info, file = "inter_info.csv")
print(paste("inter_info saved in ", getwd()))

q()